uaclient/log.py

```diff
@@ -9,3 +12,53 @@ class RedactionFilter(logging.Filter):
     def filter(self, record: logging.LogRecord):
         record.msg = util.redact_sensitive_logs(str(record.msg))
         return True
+
+
+class JsonArrayFormatter(logging.Formatter):
+    """Json Array Formatter for our logging mechanism
+    Custom made for Pro logging needs
+    """
+
+    default_time_format = "%Y-%m-%dT%H:%M:%S%z"
+    required_fields = (
+        "asctime",
+        "name",
+        "funcName",
+        "lineno",
+        "levelname",
+        "message",
+        "extra",
+    )
+
+    def format(self, record: logging.LogRecord) -> str:
+        # format the time if is  present
+        record.message = record.getMessage()
+        record.asctime = self.formatTime(record, self.default_time_format)
+
+        extra_message_dict = {}  # type: Dict[str, Any]
+        if record.exc_info:
+            extra_message_dict["exc_info"] = self.formatException(
+                record.exc_info
+            )
+        if not extra_message_dict.get("exc_info") and record.exc_text:
+            extra_message_dict["exc_info"] = record.exc_text
+        if record.stack_info:
+            extra_message_dict["stack_info"] = self.formatStack(
+                record.stack_info
+            )
+
+        # is ordered to maintain order of fields in log output
+        local_log_record = OrderedDict()  # type: Dict[str, Any]
+        # update the required fields in the order stated
+        for field in self.required_fields:
+            value = record.__dict__.get(field)
+            if field != "extra":
+                local_log_record[field] = value
+            else:
+                # Save extra, non-exception messages
+                # in another object
+                if value:
+                    extra_message_dict[field] = value
+
+        local_log_record.update({"extra_info": extra_message_dict})
```

Comment:
Since this is adding a single field, we don't need to use `update`
```suggestion
        local_log_record["extra"] = extra_message_dict
```